src/rollup/index.ts

```diff
@@ -143,133 +159,87 @@ export default function rollup(
-									plugin.ongenerate(
-										Object.assign(
-											{
-												bundle: result
-											},
-											outputOptions
-										),
-										output
-									);
-								}
-							});
-
-							return output;
-						});
-
-					Object.defineProperty(promise, 'code', throwAsyncGenerateError);
-					Object.defineProperty(promise, 'map', throwAsyncGenerateError);
-
-					return promise;
-				}
+			return applyOnbuildstartHook(inputOptions, graph)
+				.then(() => graph.buildSingle(inputOptions.input))
+				.then(chunk => {
+					return applyOnbuildendHook(graph).then(() => chunk);
+				})
+				.then(chunk => {
+					timeEnd('BUILD', 1);
 
-				const result: Bundle = {
-					imports,
-					exports,
-					modules,
+					const imports = chunk.getImportIds();
+					const exports = chunk.getExportNames();
+					const modules = graph.getCache().modules;
 
-					cache: { modules },
+					function generate(rawOutputOptions: GenericConfigObject) {
+						const outputOptions = normalizeOutputOptions(inputOptions, rawOutputOptions);
 
-					generate,
-					write: (outputOptions: OutputOptions) => {
-						if (!outputOptions || (!outputOptions.file && !outputOptions.dest)) {
+						if (outputOptions.entryNames || outputOptions.chunkNames)
 							error({
-								code: 'MISSING_OPTION',
-								message: 'You must specify output.file when doing a single-file input build'
+								code: 'INVALID_OPTION',
+								message:
+									'"entryNames" and "chunkNames" options are only supported for code-splitting builds.'
 							});
-						}
-
-						return generate(outputOptions).then(result => {
-							const file = outputOptions.file;
-							let { code, map } = result;
 
-							const promises = [];
-
-							if (outputOptions.sourcemap) {
-								let url;
-
-								if (outputOptions.sourcemap === 'inline') {
-									url = map.toUrl();
-								} else {
-									url = `${basename(file)}.map`;
-									promises.push(writeFile(file + '.map', map.toString()));
-								}
-
-								code += `//# ${SOURCEMAPPING_URL}=${url}\n`;
-							}
-
-							promises.push(writeFile(file, code));
-							return (
-								Promise.all(promises)
-									.then(() =>
-										mapSequence(graph.plugins.filter(plugin => plugin.onwrite), (plugin: Plugin) =>
-											Promise.resolve(
-												plugin.onwrite(
-													Object.assign(
-														{
-															bundle: result
-														},
-														outputOptions
-													),
-													result
-												)
+						timeStart('GENERATE', 1);
+
+						return createAddons(graph, outputOptions)
+							.then(addons => {
+								chunk.generateInternalExports(outputOptions);
+								chunk.preRender(outputOptions);
+								return chunk.render(outputOptions, addons);
+							})
+							.then(rendered => {
+								timeEnd('GENERATE', 1);
+
+								const output = {
+									name: relative(process.cwd(), resolve(outputOptions.file || inputOptions.input)),
```

Comment:
I can't remember if process is ok to use in the browser or not, would be good to check this.

Reply:
It's not, accessing `process` throws an error 😉. But then I am not exactly sure what it should return in a browser...

Reply:
Thanks, I've added a detection here. Yes the name becomes somewhat guesswork on the edge cases...